server/src/main/java/org/opensearch/index/compositeindex/datacube/startree/builder/OffHeapStarTreeBuilder.java

```diff
@@ -0,0 +1,896 @@
+        for (int i = 0; i < numStarTreeDocs; i++) {
+            starTreeDocuments.add(getStarTreeDocument(i));
+        }
+        return starTreeDocuments;
+    }
+
+    @Override
+    public Long getDimensionValue(int docId, int dimensionId) throws IOException {
+        ensureDocumentReadable(docId);
+        return starTreeDocsFileRandomInput.readLong((starTreeDocumentOffsets.get(docId) + ((long) dimensionId * Long.BYTES)));
+    }
+
+    /**
+     * Sorts and aggregates all the documents of the segment based on dimension and metrics configuration
+     *
+     * @param dimensionReaders List of docValues readers to read dimensions from the segment
+     * @param metricReaders    List of docValues readers to read metrics from the segment
+     * @return Iterator of star-tree documents
+     */
+    @Override
+    public Iterator<StarTreeDocument> sortAndAggregateSegmentDocuments(
+        SequentialDocValuesIterator[] dimensionReaders,
+        List<SequentialDocValuesIterator> metricReaders
+    ) throws IOException {
+        // Write all dimensions for segment documents into the buffer, and sort all documents using an int
+        // array
+        int documentBytesLength = 0;
+        int[] sortedDocIds = new int[totalSegmentDocs];
+        for (int i = 0; i < totalSegmentDocs; i++) {
+            sortedDocIds[i] = i;
+        }
+
+        try {
+            for (int i = 0; i < totalSegmentDocs; i++) {
+                StarTreeDocument document = getSegmentStarTreeDocument(i, dimensionReaders, metricReaders);
+                documentBytesLength = writeStarTreeDocument(document, segmentDocsFileOutput, false);
+            }
+        } finally {
+            segmentDocsFileOutput.close();
+        }
+
+        // Create an iterator for aggregated documents
+        return sortAndReduceDocuments(sortedDocIds, totalSegmentDocs, documentBytesLength);
+    }
+
+    /**
+     * Generates a star-tree for a given star-node
+     *
+     * @param startDocId  Start document id in the star-tree
+     * @param endDocId    End document id (exclusive) in the star-tree
+     * @param dimensionId Dimension id of the star-node
+     * @return iterator for star-tree documents of star-node
+     * @throws IOException throws when unable to generate star-tree for star-node
+     */
+    @Override
+    public Iterator<StarTreeDocument> generateStarTreeDocumentsForStarNode(int startDocId, int endDocId, int dimensionId)
+        throws IOException {
+        // End doc id is not inclusive but start doc is inclusive
+        // Hence we need to check if buffer is readable till endDocId - 1
+        ensureDocumentReadable(endDocId - 1);
+
+        // Sort all documents using an int array
+        int numDocs = endDocId - startDocId;
+        int[] sortedDocIds = new int[numDocs];
+        for (int i = 0; i < numDocs; i++) {
+            sortedDocIds[i] = startDocId + i;
+        }
+        new IntroSorter() {
```

Comment:
This seems like a lot of duplicate code for segment docs and reduced docs. Why have we not reused it? Lets discuss if there is some scope to reuse this logic.

Reply:
I thought about this but setPivot and comparePivot are completely different here. 

![image](https://github.com/user-attachments/assets/84ab8231-bade-43e5-9f49-406b858da978)

Reply:
Isn't the only difference in index input, and offset provider, and the actual comparison logic and datatype same? 

Can we not provide them as functions to a common sorted implementation?

Reply:
i've changed this